main/EDInitMod.F90

```diff
@@ -487,171 +662,209 @@ subroutine init_cohorts( site_in, patch_in, bc_in)
-             
-             m_struct = c_struct
-             m_leaf   = c_leaf
-             m_fnrt   = c_fnrt
-             m_sapw   = c_sapw
-             m_store  = c_store
-             m_repro  = 0._r8
-             
-          case(nitrogen_element)
-             
-             m_struct = c_struct*prt_params%nitr_stoich_p2(pft,struct_organ)
-             m_leaf   = c_leaf*prt_params%nitr_stoich_p2(pft,leaf_organ)
-             m_fnrt   = c_fnrt*prt_params%nitr_stoich_p2(pft,fnrt_organ)
-             m_sapw   = c_sapw*prt_params%nitr_stoich_p2(pft,sapw_organ)
-             m_store  = c_store*prt_params%nitr_stoich_p2(pft,store_organ)
-             m_repro  = 0._r8
-             
-          case(phosphorus_element)
-
-             m_struct = c_struct*prt_params%phos_stoich_p2(pft,struct_organ)
-             m_leaf   = c_leaf*prt_params%phos_stoich_p2(pft,leaf_organ)
-             m_fnrt   = c_fnrt*prt_params%phos_stoich_p2(pft,fnrt_organ)
-             m_sapw   = c_sapw*prt_params%phos_stoich_p2(pft,sapw_organ)
-             m_store  = c_store*prt_params%phos_stoich_p2(pft,store_organ)
-             m_repro  = 0._r8
-          end select
-
-          select case(hlm_parteh_mode)
-          case (prt_carbon_allom_hyp,prt_cnp_flex_allom_hyp )
-
-             ! Put all of the leaf mass into the first bin
-             call SetState(prt_obj,leaf_organ, element_id,m_leaf,1)
-             do iage = 2,nleafage
-                call SetState(prt_obj,leaf_organ, element_id,0._r8,iage)
+    end do
+    do pft =  1,numpft
+       if(use_pft_local(pft).eq.itrue)then
+          if(EDPftvarcon_inst%initd(pft)>nearzero) then
+
+             allocate(temp_cohort) ! temporary cohort
+
+             temp_cohort%pft         = pft
+             temp_cohort%n           = EDPftvarcon_inst%initd(pft) * patch_in%area
+             if(hlm_use_nocomp.eq.itrue)then !in nocomp mode we only have one PFT per patch
+                ! as opposed to numpft's. So we should up the initial density
+                ! to compensate (otherwise runs are very hard to compare)
+                ! this multiplies it by the number of PFTs there would have been in
+                ! the single shared patch in competition mode.          
+                ! n.b. that this is the same as currentcohort%n = %initd(pft) &AREA
+                temp_cohort%n           =  temp_cohort%n * sum(site_in%use_this_pft)
+             endif
+
+             temp_cohort%canopy_trim = 1.0_r8
+
+             !  h,dbh,leafc,n from SP values or from small initial size.
+
+             if(hlm_use_sp.eq.itrue)then
+                init = itrue
+                ! At this point, we do not know the bc_in values of tlai tsai and htop, 
+                ! so this is initializing to an arbitrary value for the very first timestep. 
+                ! Not sure if there's a way around this or not.
+                call assign_cohort_SP_properties(temp_cohort, 0.5_r8,0.2_r8, 0.1_r8,patch_in%area,init,c_leaf)
```

Comment:
This is indeed problematic. My test run shows that the arbitrary value of LAI and SAI at the first step, will give rise to balance error ("ERROR: ERROR in BalanceCheckMod.F90 at line 654") in the later time steps when the new LAI and SAI are read from satellite phenology. My solution to the initialization of SP mode is to add  satellite phenology in "clm_initializeMod.F90" before calling FATES initialization. Also, we need to add subroutine "satellite_phenology" into subroutine "wrap_update_hlmfates_dyn" in "CLMFatesInterfaceMod.F90". Then, FATES will be initializaed with proper values of LAI and SAI at the very beginning.